// @cacheable/memory/dist/index.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var index_exports = {};
__export(index_exports, {
  CacheableMemory: () => CacheableMemory,
  HashAlgorithm: () => import_utils2.HashAlgorithm,
  KeyvCacheableMemory: () => KeyvCacheableMemory,
  createKeyv: () => createKeyv,
  defaultStoreHashSize: () => defaultStoreHashSize,
  hash: () => import_utils2.hash,
  hashToNumber: () => import_utils2.hashToNumber,
  maximumMapSize: () => maximumMapSize
});
module.exports = __toCommonJS(index_exports);
var import_utils = require("@cacheable/utils");
var import_hookified = require("hookified");

// src/memory-lru.ts
var ListNode = class {
  value;
  prev = void 0;
  next = void 0;
  constructor(value) {
    this.value = value;
  }
};
var DoublyLinkedList = class {
  head = void 0;
  tail = void 0;
  nodesMap = /* @__PURE__ */ new Map();
  // Add a new node to the front (most recently used)
  addToFront(value) {
    const newNode = new ListNode(value);
    if (this.head) {
      newNode.next = this.head;
      this.head.prev = newNode;
      this.head = newNode;
    } else {
      this.head = this.tail = newNode;
    }
    this.nodesMap.set(value, newNode);
  }
  // Move an existing node to the front (most recently used)
  moveToFront(value) {
    const node = this.nodesMap.get(value);
    if (!node || this.head === node) {
      return;
    }
    if (node.prev) {
      node.prev.next = node.next;
    }
    if (node.next) {
      node.next.prev = node.prev;
    }
    if (node === this.tail) {
      this.tail = node.prev;
    }
    node.prev = void 0;
    node.next = this.head;
    if (this.head) {
      this.head.prev = node;
    }
    this.head = node;
    this.tail ??= node;
  }
  // Get the oldest node (tail)
  getOldest() {
    return this.tail ? this.tail.value : void 0;
  }
  // Remove the oldest node (tail)
  removeOldest() {
    if (!this.tail) {
      return void 0;
    }
    const oldValue = this.tail.value;
    if (this.tail.prev) {
      this.tail = this.tail.prev;
      this.tail.next = void 0;
    } else {
      this.head = this.tail = void 0;
    }
    this.nodesMap.delete(oldValue);
    return oldValue;
  }
  // Remove a specific node by value
  remove(value) {
    const node = this.nodesMap.get(value);
    if (!node) {
      return false;
    }
    if (node.prev) {
      node.prev.next = node.next;
    } else {
      this.head = node.next;
      if (this.head) {
        this.head.prev = void 0;
      }
    }
    if (node.next) {
      node.next.prev = node.prev;
    } else {
      this.tail = node.prev;
      if (this.tail) {
        this.tail.next = void 0;
      }
    }
    this.nodesMap.delete(value);
    return true;
  }
  get size() {
    return this.nodesMap.size;
  }
};

// src/index.ts
var import_utils2 = require("@cacheable/utils");

// src/keyv-memory.ts
var import_keyv = require("keyv");
var KeyvCacheableMemory = class {
  opts = {
    ttl: 0,
    useClone: true,
    lruSize: 0,
    checkInterval: 0
  }
